File: app/src/main/java/com/cgutman/adblib/TcpChannel.java
package com.cgutman.adblib;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

public class TcpChannel implements AdbChannel {
    private Socket socket;
    private InputStream is;
    private OutputStream os;

    public TcpChannel(Socket socket) throws IOException {
        this.socket = socket;
        this.is = socket.getInputStream();
        this.os = socket.getOutputStream();
    }

    @Override
    public void read(byte[] buffer) throws IOException {
        int totalRead = 0;
        while (totalRead < buffer.length) {
            int res = is.read(buffer, totalRead, buffer.length - totalRead);
            if (res < 0) throw new IOException("Conexão fechada");
            totalRead += res;
        }
    }

    @Override
    public void write(byte[] data) throws IOException {
        os.write(data);
        os.flush();
    }

    @Override
    public void close() throws IOException {
        socket.close();
    }
}

File: app/src/main/java/com/rescuedroid/rescuedroid/adb/TcpChannel.kt
package com.rescuedroid.rescuedroid.adb

import com.cgutman.adblib.AdbChannel
import java.io.InputStream
import java.io.OutputStream
import java.net.Socket

class TcpChannel(private val socket: Socket) : AdbChannel {
    private val inputStream: InputStream = socket.getInputStream()
    private val outputStream: OutputStream = socket.getOutputStream()

    override fun read(buffer: ByteArray) {
        var totalRead = 0
        while (totalRead < buffer.size) {
            val res = inputStream.read(buffer, totalRead, buffer.size - totalRead)
            if (res < 0) throw Exception("Stream closed")
            totalRead += res
        }
    }

    override fun write(buffer: ByteArray) {
        outputStream.write(buffer)
        outputStream.flush()
    }

    override fun close() {
        socket.close()
    }
}
